### PR TITLE
Improve Azure instance metadata handling

### LINKED ISSUE #70246: Azure: Instance metadata fixes and improvement
- Azure cloud provider does not correctly handle failed calls to instance metadata endpoint. We currently handle transport errors but we don't handle response errors. Ref: https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/azure/azure_instance_metadata.go#L113

we should handle `resp.statusCode != 200`-- This currently yields into empty ip address assigned to node objects.

- Remove the fall back to ARM once the above has been fixed. This removes the need for having identities in `azure.json` in case of `useInstanceMetadata = true`

- Add `user-agent: golang/kubernetes` request header for all calls to instance metadata   

- Cache the instance metadadata response and reuse for responding to `NodeAddress` and the rest of the calls. We can limit the call to `once per min` or even less.

@kubernetes/sig-azure 
@feiskyer 

**Environment**:
- Kubernetes version (use `kubectl version`): `all`
- Cloud provider or hardware configuration: `Azure`
- OS (e.g. from /etc/os-release): `Ubuntu xenial`
- Kernel (e.g. `uname -a`): `all`
- Install tools: `all`
- Others:

/king bug

### PR DESCRIPTION
**Pull request by feiskyer**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
/kind bug

**What this PR does / why we need it**:

Improve Azure instance metadata handling, which includes

* Add caches to Azure Instance Metadata
* Add user-agent: "golang/kubernetes-cloud-provider"
* Remove the fall back to ARM with empty IP
* Remove call to "v1/InstanceInfo/FD"

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #70246

**Special notes for your reviewer**:

Should cherry pick to all old releases.

**Does this PR introduce a user-facing change?**:

```release-note
Improve Azure instance metadata handling by adding caches.
```

/sig azure
/kind bug

/cc @khenidak @brendandburns @andyzhangx

### CODE REVIEW

File: pkg/cloudprovider/providers/azure/azure_instances.go [hunk 1]
```diff
@@ -82,30 +87,38 @@ func (az *Cloud) NodeAddresses(ctx context.Context, name types.NodeName) ([]v1.N
...
+				Address: address.PrivateIP,
+			})
+			if len(address.PublicIP) > 0 {
+				addresses = append(addresses, v1.NodeAddress{
+					Type:    v1.NodeExternalIP,
+					Address: address.PublicIP,
+				})
+			}
+		}
+		for _, address := range ipAddress.IPV6.IPAddress {
```
Comment (khenidak) [inline pkg/cloudprovider/providers/azure/azure_instances.go]: thank you for this @feiskyer :-)

**Review by khenidak — APPROVED** [review]
/lgtm

File: pkg/cloudprovider/providers/azure/azure_instances.go (hunk 1 above)
Comment (feiskyer) [reply pkg/cloudprovider/providers/azure/azure_instances.go]: @khenidak Do you know anyone using IPV6 now?

**Comment by titilambert** [comment]
Will you backport this commit to 1.11.5 ?

**Comment by feiskyer** [comment]
@titilambert Yep, see https://github.com/kubernetes/kubernetes/pull/70400.

**Pull request merged**